local-interchain/interchain/handlers/actions.go

```diff
@@ -0,0 +1,157 @@
+	if err != nil {
+		util.WriteError(w, err)
+		return
+	}
+
+	action := ah.Action
+	if action == "kill-all" {
+		a.killAll()
+		return
+	}
+
+	chainId := ah.ChainId
+	if _, ok := a.vals[chainId]; !ok {
+		util.Write(w, []byte(fmt.Sprintf(`{"error":"chain_id '%s' not found. Chains %v"}`, chainId, a.vals[chainId])))
+		return
+	}
+
+	ah.Cmd = strings.ReplaceAll(ah.Cmd, "%RPC%", fmt.Sprintf("tcp://%s:26657", a.vals[chainId].HostName()))
+	ah.Cmd = strings.ReplaceAll(ah.Cmd, "%CHAIN_ID%", ah.ChainId)
+	ah.Cmd = strings.ReplaceAll(ah.Cmd, "%HOME%", a.vals[chainId].HomeDir())
+
+	cmd := strings.Split(ah.Cmd, " ")
+
+	// Output can only ever be 1 thing. So we check which is set, then se the output to the user.
+	var output []byte
+	var stdout, stderr []byte
+
+	// Node / Docker Linux Actions
+	switch action {
+	case "q", "query":
+		stdout, stderr, err = (a.vals[chainId]).ExecQuery(a.ctx, cmd...)
+	case "b", "bin", "binary":
+		stdout, stderr, err = (a.vals[chainId]).ExecBin(a.ctx, cmd...)
+	case "e", "exec", "execute":
+		stdout, stderr, err = (a.vals[chainId]).Exec(a.ctx, cmd, []string{})
+	}
+
+	// Relayer Actions if the above is not used.
+	if len(stdout) == 0 && len(stderr) == 0 && err == nil {
+
+		if err := a.relayerCheck(w, r); err != nil {
+			return
+		}
+
+		switch action {
+		case "stop-relayer", "stop_relayer", "stopRelayer":
+			err = (*a.relayer).StopRelayer(a.ctx, a.eRep)
+
+		case "start-relayer", "start_relayer", "startRelayer":
+			paths := strings.FieldsFunc(ah.Cmd, func(c rune) bool {
+				return c == ',' || c == ' '
+			})
+			err = (*a.relayer).StartRelayer(a.ctx, a.eRep, paths...)
+
+		case "relayer", "relayer-exec", "relayer_exec", "relayerExec":
+			if !strings.Contains(ah.Cmd, "--home") {
+				// does this ever change for any other relayer?
+				cmd = append(cmd, "--home", "/home/relayer")
+			}
+
+			res := (*a.relayer).Exec(a.ctx, a.eRep, cmd, []string{})
+			stdout = []byte(res.Stdout)
+			stderr = []byte(res.Stderr)
+			err = res.Err
+
+		case "get_channels", "get-channels", "getChannels":
+			res, err := (*a.relayer).GetChannels(a.ctx, a.eRep, chainId)
+			if err != nil {
+				util.WriteError(w, err)
+				return
+			}
+
+			r, err := json.Marshal(res)
+			if err != nil {
+				util.WriteError(w, err)
+				return
+			}
+			stdout = r
+		}
+	}
+
+	if len(stdout) > 0 {
+		output = stdout
+	} else if len(stderr) > 0 {
+		output = stderr
+	} else if err == nil {
+		output = []byte("{}")
+	} else {
+		output = []byte(fmt.Sprintf(`%s`, err))
+	}
+
+	// Send the response
+	util.Write(w, []byte(output))
+}
+
+func (a *actions) relayerCheck(w http.ResponseWriter, r *http.Request) error {
+	var err error = nil
+
+	if a.relayer == nil {
+		util.Write(w, []byte(`{"error":"relayer not configured for this setup"}`))
+		err = fmt.Errorf("relayer not configured for this setup")
+	}
+
+	return err
+}
+
+func (a *actions) killAll() {
+	for _, v := range a.vals {
+		v.StopContainer(a.ctx)
+	}
+	a.ic.Close()
+	a.ctx.Done()
+}
```

Comment:
I haven't dove too far into this but for me, this only kills one of the vals. Maybe context related?
Also, not to add too much complexity and mapping, but this should kill the relayers too.

Again... maybe not necessary for MVP.

Reply:
Resolved & simplified: 51530bb2de7c8f5f6bef1e3b8eab1b2f32025690

tested with juno_ibc and it killed both + relayer as expected

Reply:
Update: Seems to break on my side if ic.Build is coming up. But if you cancel before then or after it is good. Need to do some more context work here for this middle part